gslib/tests/test_rsync.py

```diff
@@ -2760,28 +2760,37 @@ def _Check2():
 
   def test_dir_to_bucket_relative_minus_x(self):
     """Test that rsync -x option works with a relative regex per the docs."""
-    bucket_uri = self.CreateBucket()
     tmpdir = self.CreateTempDir(test_files=[
         'a', 'b', 'c', ('data1', 'a.txt'), ('data1', 'ok'), ('data2', 'b.txt')
     ])
 
     # Use @Retry as hedge against bucket listing eventual consistency.
     @Retry(AssertionError, tries=3, timeout_secs=1)
-    def _Check1():
+    def _check_exclude_regex(regex, want):
       """Tests rsync skips the excluded pattern."""
-      regex = 'data.[/\\\\].*\\.txt$'
+      bucket_uri = self.CreateBucket()
       # Add a trailing slash to the source directory to ensure its removed.
       local = tmpdir + ('\\' if IS_WINDOWS else '/')
       self.RunGsUtil(['rsync', '-r', '-x', regex, local, suri(bucket_uri)])
 
-      listing1 = TailSet(tmpdir, self.FlatListDir(tmpdir))
-      listing2 = TailSet(suri(bucket_uri), self.FlatListBucket(bucket_uri))
+      listing = TailSet(tmpdir, self.FlatListDir(tmpdir))
       self.assertEquals(
-          listing1,
+          listing,
           set(['/a', '/b', '/c', '/data1/a.txt', '/data1/ok', '/data2/b.txt']))
-      self.assertEquals(listing2, set(['/a', '/b', '/c', '/data1/ok']))
+      got = TailSet(suri(bucket_uri), self.FlatListBucket(bucket_uri))
+      self.assertEquals(got, want)
+
+    def _Check1():
+      _check_exclude_regex('data.[/\\\\].*\\.txt$',
+                           set(['/a', '/b', '/c', '/data1/ok']))
 
+    def _Check2():
+      _check_exclude_regex('^data|[bc]$', set(['/a']))
```

Comment:
I think even if you do not have any of the wildcard iterator filtering logic, this test will still pass? That's because rsync.py will always do the filtering. That made me realize that we never had a test case that can reproduce the symlink issue from the original bug. That's because as long as the files are valid, the existing logic present in rsync.py will always work. One way to verify would be to check for the `Skipping excluded path` string in the logs to ensure that the directory that is excluded never gets expanded (i.e the files underneath the dirs are never acted upon). I would still remove the duplicate filtering logic though, i.e skip the filtering in rsync.py if it is FileUrl.